ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatSend.java

```diff
@@ -110,44 +112,41 @@ private void calculateOffsets(boolean enableDataPrefetch) throws IOException, Me
             readSet.doPrefetch(i, 0, readSet.sizeInBytes(i));
           }
         } else {
-          // read header version
           long startTime = SystemTime.getInstance().milliseconds();
-          ByteBuffer headerVersion = ByteBuffer.allocate(Version_Field_Size_In_Bytes);
-          readSet.writeTo(i, Channels.newChannel(new ByteBufferOutputStream(headerVersion)), 0,
-              Version_Field_Size_In_Bytes);
-          logger.trace("Calculate offsets, read header version time: {}",
-              SystemTime.getInstance().milliseconds() - startTime);
-
-          headerVersion.flip();
-          short version = headerVersion.getShort();
+          BufferedInputStream bufferedInputStream =
```

Comment:
Unfortunately, it looks like `MessageReadSetIndexInputStream` does not return -1 when there is no more data but instead throws an `IOException` which is against the contract. `BufferedInputStream` may be assuming that the underlying `InputStream` fulfills` its contracts faithfully so we should change `MessageReadSetIndexInputStream` to do so. Currently it looks like there are only two production use cases of `MessageReadSetIndexInputStream` which both wrap it in a `DataInputStream` so it should be safe to make the change (tests that use it may fail but can be fixed). The class itself has a test (`MessageFormatSendTest::messageReadSetIndexInputStreamTest()`) where we should add a test for the new functionality.

Reply:
Good catch. We should add this to #932. Took a quick look into the class and I see other bugs. Such as this line in `read(byte b[], int off, int len)` method which clearly has a bug:
`long sizeToRead = Math.min(len - off, messageReadSet.sizeInBytes(indexToRead) - currentOffset);`. 
We should first fix all these issues before other changes, like you said.